### PR TITLE
prov/rxm: Add support of FI_MULTI_RECV cap

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
This patch adds support of `FI_MULTI_RECV` capability for the RxM provider.

An object of `struct rxm_multi_recv_entry` structure is supervisor object that maps to a single `struct rxm_recv_entry` object.  Whenever the previous `struct rxm_recv_entry` object was consumed and then released, the new` struct rxm_recv_entry` object is created and maps to the same `struct rxm_multi_recv_entry`. This process continues until the `rxm_ep::min_multi_recv_size` value is reached.

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango Could you take a look, please? This is PoC and there is no guarantee that this's workable solution (it hasn't been tested yet -- I'll notify when it will be tested). But I just want to gather feedback from you about this implementation. If you OK with that I'll go ahead

**Comment by gladkovdmitry17** [comment]
UPD:
This patch has been tested on **fi_rdm_multi_recv** fabtest test. Now this test is passing for RxM/verbs provider.

**Comment by a-ilango** [comment]
> (Note: the fi_getinfo_test failure is known issue -- @a-ilango, right?):

RxM should pass the fi_getinfo test. You can run runfabtests.sh with the following options to ensure all required tests are passing: `-S -R -f test_configs/ofi_rxm/ofi_rxm.exclude`

**Comment by a-ilango** [comment]
The implementation looks reasonable. Btw, is it necessary to have a separate multi recv entry? Could we just re-use the existing recv_entry by overwriting its iov info (assuming we don't have to maintain any state)?

File: prov/rxm/src/rxm_cq.c [hunk 1]
```diff
@@ -225,6 +168,143 @@ static int rxm_match_rma_iov(struct rxm_recv_entry *recv_entry,
...
+	if (rx_buf->recv_entry->flags & FI_COMPLETION) {
+		FI_DBG(&rxm_prov, FI_LOG_CQ, "writing recv completion\n");
+		ret = ofi_cq_write(rx_buf->ep->util_ep.rx_cq,
+				   rx_buf->recv_entry->context,
+				   rx_buf->recv_entry->comp_flags |
+				   (rx_buf->recv_entry->multi_recv_entry ?
+				    FI_MULTI_RECV : 0) |
+				   ((rx_buf->pkt.hdr.flags & OFI_REMOTE_CQ_DATA) ?
+				    FI_REMOTE_CQ_DATA : 0),
+				   rx_buf->pkt.hdr.size, NULL,
```
Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: The buf arg should point to the starting location of the recv buffer.

Comment (gladkovdmitry17) [reply prov/rxm/src/rxm_cq.c]: Done

**Comment by gladkovdmitry17** [comment]
> Btw, is it necessary to have a separate multi recv entry? Could we just re-use the existing recv_entry by overwriting its iov info (assuming we don't have to maintain any state)?

Yes, it's nice idea. Having a separate multi recv entry is unnecessary and can be drop to ensure best performance. Thanks!

The PR has been updated to apply comments.The PR is ready for the next round of review

File: prov/rxm/src/rxm_cq.c [hunk 2]
```diff
@@ -225,6 +169,116 @@ static int rxm_match_rma_iov(struct rxm_recv_entry *recv_entry,
...
+	int ret;
+
+	if (rx_buf->recv_entry->flags & FI_COMPLETION) {
+		FI_DBG(&rxm_prov, FI_LOG_CQ, "writing recv completion\n");
+		ret = ofi_cq_write(rx_buf->ep->util_ep.rx_cq,
+				   rx_buf->recv_entry->context,
+				   rx_buf->recv_entry->comp_flags |
+				   (rx_buf->recv_entry->flags & FI_MULTI_RECV) |
+				   ((rx_buf->pkt.hdr.flags & OFI_REMOTE_CQ_DATA) ?
+				    FI_REMOTE_CQ_DATA : 0),
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: This set of checks isn't intuitive.  Why isn't FI_REMOTE_CQ_DATA set based solely on hdr.flags & OFI_REMOTE_CQ_DATA?  I see the previous code did a bit-wise OR in of comp_flags, which I don't understand.  But OR'ing in FI_MULTI_RECV flag makes less sense to me.

Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: > Why isn't FI_REMOTE_CQ_DATA set based solely on hdr.flags & OFI_REMOTE_CQ_DATA?

The code does set it solely based on the hdr.flags

> I see the previous code did a bit-wise OR in of comp_flags, which I don't understand.

comp_flags has info on type of completion (e.g. `FI_RECV | FI_TAGGED`).  We just report all the completion flags associated with this completion and that includes FI_MULTI_RECV.

File: prov/rxm/src/rxm_cq.c [hunk 3]
```diff
@@ -225,6 +169,116 @@ static int rxm_match_rma_iov(struct rxm_recv_entry *recv_entry,
...
+				FI_WARN(&rxm_prov, FI_LOG_CQ,
+					"Unable to handle recv data\n");
+				goto done;
+			}
+		}
+
+		RXM_DBG_ADDR_TAG(FI_LOG_EP_DATA, "Enqueuing recv",
+				 rx_buf->recv_entry->addr, rx_buf->recv_entry->tag);
+		dlist_insert_tail(&rx_buf->recv_entry->entry, &rx_buf->ep->recv_queue.recv_list);
+		fastlock_release(&rx_buf->ep->recv_queue.lock);
```
Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: The above code looks similar to the one in rxm_ep_recv_common. Can it be shared in a single function?

File: prov/rxm/src/rxm_ep.c [hunk 4]
```diff
@@ -251,18 +247,15 @@ static int rxm_recv_queue_init(struct rxm_recv_queue *recv_queue, size_t size,
...
+			recv_queue->fs->buf[i].comp_flags = FI_MSG | FI_RECV;
 	} else {
 		recv_queue->match_recv = rxm_match_recv_entry_tagged;
 		recv_queue->match_unexp = rxm_match_unexp_msg_tagged;
-		for (i = recv_queue->fs->size - 1; i >= 0; i--) {
-			recv_entry = &recv_queue->fs->buf[i];
-			recv_entry->comp_flags = FI_TAGGED | FI_RECV;
-		}
+		for (i = recv_queue->fs->size - 1; i >= 0; i--)
+			recv_queue->fs->buf[i].comp_flags = FI_TAGGED | FI_RECV;
```
Comment (a-ilango) [inline prov/rxm/src/rxm_ep.c]: Like the above change, can you please move the ones unrelated to multi recv to a separate commit?

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango Can I ask you to do the next round of review?

File: prov/rxm/src/rxm_cq.c (hunk 2 above)
Comment (shefty) [reply prov/rxm/src/rxm_cq.c]: got it - I'll learn how to read code someday

File: prov/rxm/src/rxm_ep.c (hunk 4 above)
Comment (a-ilango) [reply prov/rxm/src/rxm_ep.c]: Did you forget to push the changes? I still see only one commit.

**Comment by a-ilango** [comment]
Just one comment. I think the rest of the patch looks good. 👍

**Comment by gladkovdmitry17** [comment]
> Just one comment.

I've updated PR with corresponding changes (see 2nd commit of the PR)

>I think the rest of the patch looks good. 👍

Thanks. I'm merging this PR, because it works in fabtests and doesn't affect any other functionalities of the RxM provider except for the `FI_MULTI_RECV` itself.

**Pull request merged**